src/leap/mail/mail.py

```diff
@@ -0,0 +1,697 @@
+    mailbox, tag, the result of a given query, or just a bunch of ids for
+    master documents.
+
+    Since LEAP Mail is primarily oriented to store mail in Soledad, the default
+    (and, so far, only) implementation of the store is contained in the
+    Soledad Mail Adaptor, which is passed to every collection on creation by
+    the root Account object. If you need to use a different adaptor, change the
+    adaptor class attribute in your Account object.
+
+    Store is a reference to a particular instance of the message store (soledad
+    instance or proxy, for instance).
+    """
+
+    # TODO LIST
+    # [ ] look at IMessageSet methods
+    # [ ] make constructor with a per-instance deferredLock to use on
+    #     creation/deletion?
+    # [ ] instead of a mailbox, we could pass an arbitrary container with
+    #     pointers to different doc_ids (type: foo)
+    # [ ] To guarantee synchronicity of the documents sent together during a
+    #     sync, we could get hold of a deferredLock that inhibits
+    #     synchronization while we are updating (think more about this!)
+
+    # Account should provide an adaptor instance when creating this collection.
+    adaptor = None
+    store = None
+    messageklass = Message
+
+    def __init__(self, adaptor, store, mbox_indexer=None, mbox_wrapper=None,
+                 count=None):
+        """
+        Constructor for a MessageCollection.
+        """
+        self.adaptor = adaptor
+        self.store = store
+
+        # XXX I have to think about what to do when there is no mbox passed to
+        # the initialization. We could still get the MetaMsg by index, instead
+        # of by doc_id. See get_message_by_content_hash
+        self.mbox_indexer = mbox_indexer
+        self.mbox_wrapper = mbox_wrapper
+
+        # TODO --- review this count shit. I think it's better to patch server
+        # to accept deferreds.
+        # TODO need to initialize count here because imap server does not
+        # expect a defered for the count. caller should return the deferred for
+        # prime_count (ie, initialize) when returning the collection
+        # TODO should increment and decrement when adding/deleting.
+        # TODO recent count should also be static.
+
+        if not count:
+            count = 0
+        self._count = count
+
+    def is_mailbox_collection(self):
+        """
+        Return True if this collection represents a Mailbox.
+        :rtype: bool
+        """
+        return bool(self.mbox_wrapper)
+
+    @property
+    def mbox_name(self):
+        wrapper = getattr(self, "mbox_wrapper", None)
+        if not wrapper:
+            return None
+        return wrapper.mbox
+
+    @property
+    def mbox_uuid(self):
+        wrapper = getattr(self, "mbox_wrapper", None)
+        if not wrapper:
+            return None
+        return wrapper.uuid
+
+    def get_mbox_attr(self, attr):
+        return getattr(self.mbox_wrapper, attr)
```

Comment:
good point. I've changed it to raise an exception if the method is used with a collection that is not a mailbox collection.
This makes me think that this class is badly designed, and that the MailboxCollection could be better implemented by subclassing... what do you think?